File: packages/b2c-tooling-sdk/src/operations/jobs/ocapi-mapping.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Mapping helpers from raw OCAPI shapes (snake_case) to canonical
 * {@link JobExecutionInfo} (camelCase).
 *
 * These exist as transitional utilities to bridge the two API shapes. They
 * will be deleted along with the OCAPI ops once OCAPI is removed.
 *
 * @module operations/jobs/ocapi-mapping
 */
import type {JobExecution, JobStepExecution} from './run.js';
import type {JobExecutionInfo, JobExecutionSearchResults, JobStepExecutionResult} from './types.js';

function mapStepExecution(step: JobStepExecution): JobStepExecutionResult {
  return {
    id: step.id,
    stepId: step.step_id,
    executionStatus: step.execution_status,
    exitStatus: step.exit_status
      ? {
          code: step.exit_status.code ?? '',
          message: step.exit_status.message,
          status: step.exit_status.status as 'ok' | 'error' | undefined,
        }
      : undefined,
    duration: step.duration,
  };
}

/** Map a raw OCAPI {@link JobExecution} into the canonical shape. */
export function mapOcapiExecution(ocapi: JobExecution): JobExecutionInfo {
  return {
    id: ocapi.id ?? '',
    jobId: ocapi.job_id ?? '',
    executionStatus: (ocapi.execution_status ?? 'unknown') as JobExecutionInfo['executionStatus'],
    exitStatus: ocapi.exit_status
      ? {
          code: ocapi.exit_status.code ?? '',
          message: ocapi.exit_status.message,
          status: ocapi.exit_status.status as 'ok' | 'error' | undefined,
        }
      : undefined,
    startTime: ocapi.start_time,
    endTime: ocapi.end_time,
    duration: ocapi.duration,
    stepExecutions: ocapi.step_executions?.map(mapStepExecution),
    logFilePath: ocapi.log_file_path,
    isLogFileExisting: ocapi.is_log_file_existing,
    parameters: ocapi.parameters,
    _raw: ocapi,
  };
}

function mapCanonicalStepExecution(step: JobStepExecutionResult): JobStepExecution {
  return {
    id: step.id,
    step_id: step.stepId,
    execution_status: step.executionStatus as JobStepExecution['execution_status'],
    exit_status: step.exitStatus
      ? {
          code: step.exitStatus.code,
          message: step.exitStatus.message,
          status: step.exitStatus.status,
        }
      : undefined,
    duration: step.duration,
  } as JobStepExecution;
}

/**
 * Map a canonical {@link JobExecutionInfo} back into the raw OCAPI
 * {@link JobExecution} (snake_case) shape.
 *
 * The reverse of {@link mapOcapiExecution}. System-job operations
 * (site-archive import/export, CAP install/uninstall) expose the raw OCAPI
 * `JobExecution` in their public result/error types; when those operations are
 * served over SCAPI, the canonical result is mapped back through this so the
 * public contract stays identical across backends.
 *
 * Prefers the original OCAPI payload when present in `_raw` (lossless
 * round-trip for the OCAPI path); otherwise projects the canonical fields.
 */
export function mapCanonicalToOcapiExecution(canonical: JobExecutionInfo): JobExecution {
  if (canonical._raw && typeof canonical._raw === 'object' && 'execution_status' in canonical._raw) {
    return canonical._raw as JobExecution;
  }

  return {
    id: canonical.id,
    job_id: canonical.jobId,
    execution_status: canonical.executionStatus as JobExecution['execution_status'],
    exit_status: canonical.exitStatus
      ? {
          code: canonical.exitStatus.code,
          message: canonical.exitStatus.message,
          status: canonical.exitStatus.status,
        }
      : undefined,
    start_time: canonical.startTime,
    end_time: canonical.endTime,
    duration: canonical.duration,
    step_executions: canonical.stepExecutions?.map(mapCanonicalStepExecution),
    log_file_path: canonical.logFilePath,
    is_log_file_existing: canonical.isLogFileExisting,
    parameters: canonical.parameters,
  } as JobExecution;
}

/** Map a raw OCAPI search result into the canonical shape. */
export function mapOcapiSearchResult(result: {
  total: number;
  count: number;
  start: number;
  hits: JobExecution[];
}): JobExecutionSearchResults {
  return {
    total: result.total,
    limit: result.count,
    offset: result.start,
    hits: result.hits.map(mapOcapiExecution),
  };
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/jobs/ocapi_mapping.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Mapping helpers between raw OCAPI shapes (snake_case) and canonical
:class:`JobExecutionInfo` (camelCase-ish).

Mirrors ``src/operations/jobs/ocapi-mapping.ts``. Transitional utilities to
bridge the two API shapes; deleted along with the OCAPI ops once OCAPI is
removed.
"""

from __future__ import annotations

from typing import Any

from b2c_tooling_sdk.operations.jobs.run import JobExecution, JobExecutionSearchResult, JobStepExecution
from b2c_tooling_sdk.operations.jobs.types import (
    JobExecutionInfo,
    JobExecutionSearchResults,
    JobExitStatus,
    JobStepExecutionResult,
)


def _map_exit_status(exit_status: dict[str, Any] | None) -> JobExitStatus | None:
    if not exit_status:
        return None
    return JobExitStatus(
        code=exit_status.get("code") or "",
        message=exit_status.get("message"),
        status=exit_status.get("status"),
    )


def _map_step_execution(step: dict[str, Any]) -> JobStepExecutionResult:
    return JobStepExecutionResult(
        id=step.get("id"),
        step_id=step.get("step_id"),
        execution_status=step.get("execution_status"),
        exit_status=_map_exit_status(step.get("exit_status")),
        duration=step.get("duration"),
    )


def map_ocapi_execution(ocapi: JobExecution) -> JobExecutionInfo:
    """Map a raw OCAPI :data:`JobExecution` into the canonical shape."""
    step_executions = ocapi.get("step_executions")
    return JobExecutionInfo(
        id=ocapi.get("id") or "",
        job_id=ocapi.get("job_id") or "",
        execution_status=ocapi.get("execution_status") or "unknown",
        exit_status=_map_exit_status(ocapi.get("exit_status")),
        start_time=ocapi.get("start_time"),
        end_time=ocapi.get("end_time"),
        duration=ocapi.get("duration"),
        step_executions=[_map_step_execution(s) for s in step_executions] if step_executions else None,
        log_file_path=ocapi.get("log_file_path"),
        is_log_file_existing=ocapi.get("is_log_file_existing"),
        parameters=ocapi.get("parameters"),
        raw=ocapi,
    )


def _map_canonical_step_execution(step: JobStepExecutionResult) -> JobStepExecution:
    result: JobStepExecution = {
        "id": step.id,
        "step_id": step.step_id,
        "execution_status": step.execution_status,
        "duration": step.duration,
    }
    if step.exit_status is not None:
        result["exit_status"] = {
            "code": step.exit_status.code,
            "message": step.exit_status.message,
            "status": step.exit_status.status,
        }
    else:
        result["exit_status"] = None
    return result


def map_canonical_to_ocapi_execution(canonical: JobExecutionInfo) -> JobExecution:
    """Map a canonical :class:`JobExecutionInfo` back into the raw OCAPI shape.

    The reverse of :func:`map_ocapi_execution`. Prefers the original OCAPI
    payload when present in ``raw`` (lossless round-trip for the OCAPI path);
    otherwise projects the canonical fields.
    """
    raw = canonical.raw
    if isinstance(raw, dict) and "execution_status" in raw:
        return raw

    exit_status = None
    if canonical.exit_status is not None:
        exit_status = {
            "code": canonical.exit_status.code,
            "message": canonical.exit_status.message,
            "status": canonical.exit_status.status,
        }

    step_executions = None
    if canonical.step_executions is not None:
        step_executions = [_map_canonical_step_execution(s) for s in canonical.step_executions]

    return {
        "id": canonical.id,
        "job_id": canonical.job_id,
        "execution_status": canonical.execution_status,
        "exit_status": exit_status,
        "start_time": canonical.start_time,
        "end_time": canonical.end_time,
        "duration": canonical.duration,
        "step_executions": step_executions,
        "log_file_path": canonical.log_file_path,
        "is_log_file_existing": canonical.is_log_file_existing,
        "parameters": canonical.parameters,
    }


def map_ocapi_search_result(result: JobExecutionSearchResult) -> JobExecutionSearchResults:
    """Map a raw OCAPI search result into the canonical shape."""
    return JobExecutionSearchResults(
        total=result.total,
        limit=result.count,
        offset=result.start,
        hits=[map_ocapi_execution(hit) for hit in result.hits],
    )


__all__ = [
    "map_canonical_to_ocapi_execution",
    "map_ocapi_execution",
    "map_ocapi_search_result",
]
